### PR TITLE
chore: allow empty statements to have no range

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
If you `metavariable-pattern` on a metavariable matching an empty sequence of statements, it will complain that an internal error happened, because there is no range info. In reality, there being no range info is actually OK in this situation.

For instance, this rule will emit an error message: https://semgrep.dev/playground/s/RelZE

## How:
This PR makes it so `AST_generic_helpers.range_of_any` returns a new variant type, with two variants `No_range_expected` and `No_range_error`. The former is for ASTs which have no real tokens, by our own mistake (for instance, by spawning too many fake tokens), and the latter is for ASTs which are expected to have no tokens, but are expected to (such as `G.Ss []`)

## Test plan:
Tested manually.

### CODE REVIEW

File: src/engine/Metavariable_pattern.ml [hunk 1]
```diff
@@ -122,13 +122,20 @@ let get_nested_metavar_pattern_bindings get_nested_formula_matches env r mvar
...
-      | None ->
-          error env
-            (Common.spf
-               "metavariable-pattern failed because we lack range info for %s, \
-                please file a bug report"
-               mvar);
-          []
+      | None -> (
+          match mval with
+          | MV.Ss [] ->
```
Comment (aryx) [inline src/engine/Metavariable_pattern.ml:129]: there are actually more cases like this. See Core_json_output.range_of_any_opt.
Maybe there is some code factorization to do, maybe a helper function in Metavariable.ml

Also would be good to add a test showing that now we don't crash on the code you've added.

**Review by aryx — CHANGES_REQUESTED** [review]

File: src/engine/Metavariable_pattern.ml (hunk 1 above)
Comment (brandonspark) [reply src/engine/Metavariable_pattern.ml:129]: i can't add a test i think, because the "crash" isn't really a crash, its just a reported error message

Comment (brandonspark) [reply src/engine/Metavariable_pattern.ml:129]: (i could add a CLI test, but i was thinking because pysemgrep is going away, that may not be worth it)

**Comment by brandonspark** [comment]
@aryx

File: src/core/Metavariable.ml [hunk 2]
```diff
@@ -160,11 +160,11 @@ let location_aware_equal_mvalue mval1 mval2 =
...
-      ( AST_generic_helpers.range_of_any_opt any1,
-        AST_generic_helpers.range_of_any_opt any2 )
+      ( AST_generic_helpers.range_of_any any1,
+        AST_generic_helpers.range_of_any any2 )
     in
     match (range1, range2) with
-    | Some (l1, r1), Some (l2, r2) ->
+    | Range (l1, r1), Range (l2, r2) ->
         Tok.equal_location l1 l2 && Tok.equal_location r1 r2
     | _ -> true
```
Comment (aryx) [inline src/core/Metavariable.ml:169]: This seems wrong ... if one is a Range, and the other a No_range_xxx, then we consider the ranges equal?

File: libs/ast_generic/AST_generic_helpers.mli [hunk 3]
```diff
@@ -1,3 +1,8 @@
+type any_range =
+  | No_range_error
```
Comment (aryx) [inline libs/ast_generic/AST_generic_helpers.mli:2]: Please add a comment explaining the difference between the 2 No_range_xxx.

**Review by aryx — APPROVED** [review]

**Pull request closed without merging**